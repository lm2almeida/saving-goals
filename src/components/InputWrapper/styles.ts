import styled from 'styled-components';

import media from '../../commons/styles/utils/media';

interface ButtonWrapperProps {
  disabled: boolean;
  isLeft: boolean;
}

const Container = styled.fieldset`
  &:not(:last-of-type) {
    margin-bottom: var(--spacing-medium);

    ${media.desktop`
      margin-bottom: 0;
      margin-right: var(--spacing-medium);
    `}
  }
`;

const InputWrapper = styled.div`
  border: var(--border-default);
  border-radius: var(--border-radius-small);
  display: flex;
  overflow: hidden;
  width: 100%;

  > * {
    border: none;
    background-color: var(--neutral-white);
  }
`;

const Label = styled.label`
  color: var(--colors-gray-900);
  display: block;
  font-size: 12px;
  margin-bottom: var(--spacing-xxsmall);

  ${media.desktop`
    font-size: 14px;
  `}
`;

const ButtonWrapper = styled.button<ButtonWrapperProps>`
  --padding-left: ${({ isLeft }) =>
    isLeft ? 'var(--spacing-small)' : 'var(--spacing-xsmall)'};
  --padding-right: ${({ isLeft }) =>
    isLeft ? 'var(--spacing-xsmall)' : 'var(--spacing-small)'};
  padding: var(--spacing-small) var(--padding-right) var(--spacing-small)
    var(--padding-left);

  svg {
    --size: 24px;
    height: var(--size);
    width: var(--size);

    path {
      --fill: ${({ disabled }) =>
        disabled
          ? 'var(--colors-blue-gray-100)'
          : 'var(--colors-blue-gray-300)'};
      fill: var(--fill) !important;
    }
  }
`;

export { Container, InputWrapper, Label, ButtonWrapper };
