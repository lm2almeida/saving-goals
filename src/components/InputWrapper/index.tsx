import React from 'react';

import * as S from './styles';

type ButtonProps = {
  onClick?: () => void;
  disabled?: boolean;
  label: string;
};

type Button = {
  component: React.ReactElement;
  props: ButtonProps;
};

export type Buttons = {
  left?: Button;
  right?: Button;
};

interface InputWrapperProps {
  children: React.ReactNode;
  id: string;
  label: string;
  buttons?: Buttons;
}

function renderActionButton(
  { component, props: { disabled = false, onClick, label } }: Button,
  isLeft = false
) {
  const handleClick = () => {
    if (!disabled && onClick) {
      onClick();
    }
  };

  return (
    <S.ButtonWrapper
      onClick={handleClick}
      disabled={disabled}
      isLeft={isLeft}
      type="button"
      title={label}
      aria-label={label}
    >
      {component}
    </S.ButtonWrapper>
  );
}

function InputWrapper({
  children,
  label,
  id,
  buttons,
}: InputWrapperProps): JSX.Element {
  return (
    <S.Container>
      <S.Label htmlFor={id}>{label}</S.Label>
      <S.InputWrapper>
        {buttons?.left && renderActionButton(buttons.left, true)}
        {children}
        {buttons?.right && renderActionButton(buttons.right)}
      </S.InputWrapper>
    </S.Container>
  );
}

export default InputWrapper;
