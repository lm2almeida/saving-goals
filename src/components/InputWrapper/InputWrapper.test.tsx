import { render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import InputWrapper, { Buttons } from '.';

function setup(buttons?: Buttons): RenderResult {
  return render(
    <InputWrapper id="test" label="test" buttons={buttons || {}}>
      <input id="test" name="test" />
    </InputWrapper>
  );
}

describe('components/InputWrapper', () => {
  it('should render component', () => {
    const { asFragment } = setup();

    expect(screen.getByText(/test/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/test/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render buttons and call onClick', () => {
    const mockLeftClick = jest.fn();
    const mockRightClick = jest.fn();
    const buttons = {
      left: {
        component: <span>click left</span>,
        props: {
          onClick: mockLeftClick,
        },
      },
      right: {
        component: <span>click right</span>,
        props: {
          onClick: mockRightClick,
        },
      },
    };

    setup(buttons);

    userEvent.click(screen.getByRole('button', { name: /click left/ }));
    userEvent.click(screen.getByRole('button', { name: /click right/ }));

    expect(mockLeftClick).toHaveBeenCalled();
    expect(mockRightClick).toHaveBeenCalled();
  });

  it('should render buttons and call not onClick if disabled or no onClick', () => {
    const mockLeftClick = jest.fn();
    const mockRightClick = jest.fn();
    const buttons = {
      left: {
        component: <span>click left</span>,
        props: {},
      },
      right: {
        component: <span>click right</span>,
        props: {
          onClick: mockRightClick,
          disabled: true,
        },
      },
    };

    setup(buttons);

    userEvent.click(screen.getByRole('button', { name: /click left/ }));
    userEvent.click(screen.getByRole('button', { name: /click right/ }));

    expect(mockLeftClick).not.toHaveBeenCalled();
    expect(mockRightClick).not.toHaveBeenCalled();
  });
});
