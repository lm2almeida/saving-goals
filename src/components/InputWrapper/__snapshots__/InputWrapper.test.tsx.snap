// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/InputWrapper should render component 1`] = `
<DocumentFragment>
  .c0:not(:last-of-type) {
  margin-bottom: var(--spacing-medium);
}

.c2 {
  border: 1px solid var(--colors-blue-gray-50);
  border-radius: var(--border-radius-small);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: hidden;
  width: 100%;
}

.c2 > * {
  border: none;
  background-color: var(--neutral-white);
}

.c1 {
  color: var(--colors-gray-900);
  display: block;
  font-size: 12px;
  margin-bottom: 4px;
}

@media screen and (min-width:1280px) {
  .c0:not(:last-of-type) {
    margin-bottom: 0;
    margin-right: var(--spacing-medium);
  }
}

@media screen and (min-width:1280px) {
  .c1 {
    font-size: 14px;
  }
}

<fieldset
    class="c0"
  >
    <label
      class="c1"
      for="test"
    >
      test
    </label>
    <div
      class="c2"
    >
      <input
        id="test"
        name="test"
      />
    </div>
  </fieldset>
</DocumentFragment>
`;
